lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -1084,11 +1085,31 @@ void FIRRTLModuleLowering::lowerModuleBody(
 //===----------------------------------------------------------------------===//
 
 namespace {
+
+struct MixedModuleNamespace : NamespaceBase {
+  MixedModuleNamespace() {}
+  MixedModuleNamespace(hw::HWModuleOp module) { add(module); }
+
+  /// Populate the namespace from a module-like operation. This namespace will
+  /// be composed of the `inner_sym`s of the module's ports and declarations.
+  void add(hw::HWModuleOp module) {
+    for (auto port : module.getAllPorts())
+      if (port.sym && !port.sym.getValue().empty())
+        internal.insert(port.sym.getValue());
+    module.walk([&](Operation *op) {
+      auto attr = op->getAttrOfType<StringAttr>("inner_sym");
+      if (attr)
+        internal.insert(attr.getValue());
+    });
+  }
+};
+
```

Comment:
This looks very useful! Maybe we should move `NamespaceBase` into the HW dialect and provide `HWModuleOp` implementations there, and then have FIRRTL add its own namespaces on top.

Reply:
Great idea. Depending on timing, I'll either do this in an NFC commit before this lands or afterwards.

Reply:
I moved `NamespaceBase` out of FIRRTL and into `circt/Support` and renamed it `Namespace` in 85b24de8cb7f5e8db0a58a7dd50310b1dde4eb7e.